File: nativeapps/render.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
    Very basic templating mechanics.

    This is used to render the application lists in each of the tabs.
"""

import os
import json

import nativeapps.io


TEMPLATE = """
     <tr>
        <th scope="row">
            <a href="%s">
                <img style="width: 2em; height: 2em;" src="static/img/%s.png"/>
            </a>
        </th>
        <td>%s</td>
        <td>
            <button class="btn btn-link"
                    data-toggle="collapse"
                    data-target="#collapseID%i">%s</button>
        </td>
    </tr>
    <tr>
        <td colspan="4">
            <div class="collapse" id="collapseID%i">
                <div class="card card-body">
                    %s
                </div>
            </div>
        </td>
    </tr>
""" # link, icon, name, version, metadata, sequence_id, sequence_id, metadata

def handle_list(sequence):
    """
        Given an iterator:
            ["hello", "world"]
        Generate:
            <p>hello</p><p>world</p>
    """
    string = ""
    for value in sequence:
        if isinstance(value, dict):
            string += handle_dict(value)
        else:
            value = str(value)
            string += "<p>%s</p>\n" % (str(value))
    return string

def handle_dict(dictionary):
    """
        Prints HTML-formatted tables.
    """
    string = """
        <table class="table table-sm table-responsive alert alert-dark">
        <thead>
            <tr>
                <th scope="col">Key</th>
                <th scope="col">Value</th>
            </tr>
        </thead>
        <tbody>
    """

    for key, value in dictionary.items():
        if isinstance(value, list) and value:
            value = handle_list(value)
        elif isinstance(value, dict):
            value = handle_dict(value)
        elif isinstance(value, int):
            value = str(value)
        elif isinstance(value, unicode):
            value = value.encode("utf-8")
        string += "<tr><td>{key}</td><td>{value}</td></tr>\n".format(key=key, value=value)

    string += "</tbody></table>"
    return string

def android(url, rootdir):
    """
        Render the HTML for APKs.
    """
    template = ""
    i = 0
    for path in nativeapps.io.ls(rootdir, r".*\.apk$"):
        directory = os.path.dirname(path)
        name, version = os.path.basename(directory).split("-", 1)
        meta = json.load(open(os.path.join(directory, "metadata.json"), "r"))
        app_url = path.decode("utf-8").replace(rootdir, url).encode("utf-8")
        template += TEMPLATE % (app_url, "android",
                                name, i, version, i, handle_dict(meta))
        i += 1
    return template

def ios(url, rootdir):
    """
        Render the HTML for IPAs.

        The two differences from android are:
            * We have to extract the name/version with a slightly different logic
            * The application's link is this weird itms-services:// thing.
    """
    template = ""
    i = 0
    url = "itms-services://?action=download-manifest&url=" + url

    for path in nativeapps.io.ls(rootdir, r".*manifest.plist$"):
        directory = os.path.dirname(path)
        name, version = os.path.basename(directory).split("-", 1)
        meta = json.load(open(os.path.join(directory, "metadata.json"), "r"))
        app_url = path.decode("utf-8").replace(rootdir, url).encode("utf-8")
        template += TEMPLATE % (app_url, "ios",
                                name, i, version, i, handle_dict(meta))
        i += 1
    return template

File: nativeapps/static/js/render.js
function load_app() {
    $.ajax({
        url: "/application"
    }).then(function(data) {
        render_applications(data.applications);
        render_deleteapps(data.applications);
        render_apptags(data.applications);
    });
}

function render_applications(applications) {
    $(`div#ipa table tbody`).empty();
    $(`div#apk table tbody`).empty();
    applications.forEach(function (app, index) {
        var type = app.type;
        var dom_obj = render_app(index, app);
        $(`div#${type} > table > tbody`).append(dom_obj);
    });
}

function render_app(id, application) {
    let link = application.link;
    let type = application.type;
    let name = application.name;
    let version = application.version;
    let metadata = JSON.parse(application.metadata);
    let metadata_dom = render_object(metadata);
    let tags = "";

    if (Array.isArray(metadata.tags)) {
        metadata.tags.forEach(function (tag_name, index) {
            tags = `${tags} <span class="badge badge-primary">${tag_name}</span>`
        });
    }

    return `
    <tr>
        <td scope="row">
            <a href="${link}">
                <img style="width: 2em; height: 2em;" src="static/img/${type}.png"/>
            </a>
        </td>
        <td>
            ${name}-${version}
            ${tags}
        </td>
        <td>
            <button class="btn btn-link"
                    data-toggle="collapse"
                    data-target="#collapseID${id}">
                <img style="width: 1.5em; height: 1.5em;" src="static/img/help.png"/>
            </button>
        </td>
    </tr>
    <tr>
        <td colspan="4">
            <div class="collapse" id="collapseID${id}">
                <div class="card card-body">
                   ${metadata_dom}
                </div>
            </div>
        </td>
    </tr>`;
}

function get_type(object) {
    let type = typeof object;
    if (type == "object"){
        if (Array.isArray(object)) {
            return "array"
        } else {
            return "object"
        }
    } else {
        return "string"
    }
}


function render_array(meta) {
    var template = `<ul class="list-unstyled">`;
    meta.forEach(function (item, index) {
        if (get_type(item) == "object") {
            template = template + render_object(item);
        } else {
            template = template + `<li>${item}</li>`;
        }
    });
    return template + `</ul>`;
}


function render_object(meta) {
    let table = `
    <table class="table small table-sm table-responsive alert alert-dark">
        <thead>
            <tr>
                <th scope="col">Key</th>
                <th scope="col">Value</th>
            </tr>
        </thead>
        <tbody>`;

    for (var key in meta) {
        var value = meta[key];
        var type = get_type(value);
        if (type == "object") {
            value = render_object(value);
        } else if (type == "array") {
            value = render_array(value);
        }

        table = table + `
            <tr>
                <td>${key}</td>
                <td>${value}</td>
            </tr>
        `;
    }

    return table + "</table>";
}


function render_deleteapps(apps) {
    $("form#deleteApplications select").empty();
    apps.forEach(function (app, index) {
        var name = app.name + "-" + app.version + "." + app.type;
        $("form#deleteApplications select").append(`<option>${name}</option>`);
    });
}

function render_apptags(apps) {
    $("form select#applicationNameTags").empty();
    apps.forEach(function (app, index) {
        var name = app.name + "-" + app.version + "." + app.type;
        $("form select#applicationNameTags").append(`<option>${name}</option>`);
    });
    $("select#applicationTags").tagsinput("removeAll", { preventPost: true });
    tags = JSON.parse(apps[0].metadata).tags
    if (tags) {
        tags.forEach(function (tag_name, index) {
            $("select#applicationTags").tagsinput("add", tag_name, { preventPost: true });
        });
    }
}

